rootfs/etc/nginx/lua/balancer/ewma.lua

```diff
@@ -123,19 +123,41 @@ local function pick_and_score(peers, k)
       lowest_score_index, lowest_score = i, new_score
     end
   end
-  return peers[lowest_score_index]
+  return peers[lowest_score_index], lowest_score
+end
+
+-- slow_start_ewma is something we use to avoid sending too many requests
+-- to the newly introduced endpoints. We currently use average ewma values
+-- of existing endpoints.
+local function calculate_slow_start_ewma(self)
+  local total_ewma = 0
+  local endpoints_count = 0
+
+  for _, endpoint in pairs(self.peers) do
+    local endpoint_string = endpoint.address .. ":" .. endpoint.port
+    local ewma = ngx.shared.balancer_ewma:get(endpoint_string)
+
+    if ewma then
+      endpoints_count = endpoints_count + 1
+      total_ewma = total_ewma + ewma
+    end
+  end
+
+  return total_ewma / endpoints_count
 end
 
 function _M.balance(self)
   local peers = self.peers
-  local endpoint = peers[1]
+  local endpoint, ewma_score = peers[1], -1
 
   if #peers > 1 then
     local k = (#peers < PICK_SET_SIZE) and #peers or PICK_SET_SIZE
     local peer_copy = util.deepcopy(peers)
-    endpoint = pick_and_score(peer_copy, k)
+    endpoint, ewma_score = pick_and_score(peer_copy, k)
   end
 
+  ngx.var.balancer_ewma_score = ewma_score
```

Comment:
Nitpic: why a var instead of ctx ?

Reply:
ctx is lua specific - this is going to be part of public ingress too, we should be able to use that variable in nginx access log